inc/admin-layouts.php

```diff
@@ -0,0 +1,313 @@
+		if( empty( $directories ) || ! is_array( $directories ) ) {
+			$directories = array();
+		}
+		
+		return $directories;
+	}
+	
+	/**
+	 * Gets all the prebuilt layouts
+	 */
+	function action_get_prebuilt_layouts() {
+		if ( empty( $_REQUEST['_panelsnonce'] ) || ! wp_verify_nonce( $_REQUEST['_panelsnonce'], 'panels_action' ) ) {
+			wp_die();
+		}
+		
+		// Get any layouts that the current user could edit.
+		header( 'content-type: application/json' );
+		
+		$type   = ! empty( $_REQUEST['type'] ) ? $_REQUEST['type'] : 'directory-siteorigin';
+		$search = ! empty( $_REQUEST['search'] ) ? trim( strtolower( $_REQUEST['search'] ) ) : '';
+		$page   = ! empty( $_REQUEST['page'] ) ? intval( $_REQUEST['page'] ) : 1;
+		
+		$return = array(
+			'title' => '',
+			'items' => array()
+		);
+		if ( $type == 'prebuilt' ) {
+			$return['title'] = __( 'Theme Defined Layouts', 'siteorigin-panels' );
+			
+			// This is for theme bundled prebuilt directories
+			$layouts = apply_filters( 'siteorigin_panels_prebuilt_layouts', array() );
+			
+			foreach ( $layouts as $id => $vals ) {
+				if ( ! empty( $search ) && strpos( strtolower( $vals['name'] ), $search ) === false ) {
+					continue;
+				}
+				
+				$return['items'][] = array(
+					'title'       => $vals['name'],
+					'id'          => $id,
+					'type'        => 'prebuilt',
+					'description' => isset( $vals['description'] ) ? $vals['description'] : '',
+					'screenshot'  => ! empty( $vals['screenshot'] ) ? $vals['screenshot'] : ''
+				);
+			}
+			
+			$return['max_num_pages'] = 1;
+		} elseif ( substr( $type, 0, 10 ) == 'directory-' ) {
+			$return['title'] = __( 'Layouts Directory', 'siteorigin-panels' );
+			
+			// This is a query of the prebuilt layout directory
+			$query = array();
+			if ( ! empty( $search ) ) {
+				$query['search'] = $search;
+			}
+			$query['page'] = $page;
+			
+			$directory_id = str_replace( 'directory-', '', $type );
+			$directories = $this->get_directories();
+			$directory = ! empty( $directories[ $directory_id ] ) ? $directories[ $directory_id ] : false;
+			
+			if( empty( $directory ) ) {
+				return false;
+			}
+			
+			$url = add_query_arg( $query, $directory[ 'url' ] . 'wp-admin/admin-ajax.php?action=query_layouts' );
+			if( ! empty( $directory[ 'args' ] ) && is_array( $directory[ 'args' ] ) ) {
+				$url = add_query_arg( $directory[ 'args' ], $url );
+			}
+			
+			$url = apply_filters( 'siteorigin_panels_layouts_directory_url', $url );
+			$response = wp_remote_get( $url );
+			
+			if ( is_array( $response ) && $response['response']['code'] == 200 ) {
+				$results = json_decode( $response['body'], true );
+				if ( ! empty( $results ) && ! empty( $results['items'] ) ) {
+					foreach ( $results['items'] as $item ) {
+						$item['id']         = $item['slug'];
+						$item['type']       = $type;
+						
+						if( empty( $item['screenshot'] ) ) {
+							$screenshot_url = add_query_arg( 'screenshot_preview', 1, $item['preview'] );
```

Comment:
The API should always be sending that. It just comes from the_permalink of the layout post type. I agree though, we should make it handle weird return values as much as possible.